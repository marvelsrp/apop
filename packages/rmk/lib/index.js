const reduceObject = (origin, remakers) => {
  return remakers.reduce((state, remake) => {
    return remake(state);
  }, origin);
};

const reduceArray = (origin, remakers) => {
  return origin.reduce((state, element) => {
    state.push(reduceObject(element, remakers));
    return state;
  }, []);
};

const reduce = (origin, reamakers) => {
  if (isArray(origin)) {
    return reduceArray(origin, mutators);
  } else if (isObject(origin)) {
    return reduceObject(origin, reamakers);
  } else {
    throw new Error('Wrong input arguments');
  }
};

const rmk = (...reamakers) => {
  return origin => reduce(origin, reamakers);
};

export default rmk;
