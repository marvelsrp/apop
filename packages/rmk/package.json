{
  "name": "@rmk/rmk",
  "version": "1.0.0",
  "scripts": {
    "prepublish": "babel -d ./ ./lib",
    "dev": "babel -d ./ ./lib --watch",
    "test": "jest --coverage",
    "test-dev": "jest --watch"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tuchk4/rmk.git"
  },
  "dependencies": {
    "@rmk/clear": "=1.0.0"
  },
  "license": "ISC",
  "main": "index.js"
}
