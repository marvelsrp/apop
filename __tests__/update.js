import rmk from '../lib';
import rmkActions from '../lib/actions';


export const before = {
  "date": new Date(2001,11,3),
};
export const after = {
  "date": new Date(2001,11,3),
  "year": 2001,
  "month": 11,
  "day": 3
};

describe('Update action', () => {
  it('Field function object', () => {
    const formula = rmk(
      rmkActions.update({
        year: localState => localState.date.getFullYear(),
        month: localState => localState.date.getMonth(),
        day: localState => localState.date.getDate(),
      })
    );
    const result = formula(before);
    expect(result).toEqual(after);
  });

  it('Field static object', () => {
    const formula = rmk(
      rmkActions.update({
        year: after.year,
        month: after.month,
        day: after.day,
      })
    );
    const result = formula(before);
    expect(result).toEqual(after);
  });

  it('Field function array', () => {
    const formula = rmk(
      rmkActions.update({
        year: localState => localState.date.getFullYear(),
        month: localState => localState.date.getMonth(),
        day: localState => localState.date.getDate(),
      })
    );
    const result = formula([before]);
    expect(result).toEqual([after]);
  });

  it('Field static array', () => {
    const additionalFields = {
      year: 2000,
      month: 1,
      day: 1,
    };
    const formula = rmk(rmkActions.update(additionalFields));
    const result = formula([before]);
    const trueResult = [{ ...after, ...additionalFields }];
    expect(result).toEqual(trueResult);
  });
});
