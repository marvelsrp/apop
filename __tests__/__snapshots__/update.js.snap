// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field function array 1`] = `
Array [
  Object {
    "date": 2001-12-02T22:00:00.000Z,
    "day": 3,
    "month": 11,
    "year": 2001,
  },
]
`;

exports[`Field function object 1`] = `
Object {
  "date": 2001-12-02T22:00:00.000Z,
  "day": 3,
  "month": 11,
  "year": 2001,
}
`;

exports[`Field static array 1`] = `
Array [
  Object {
    "date": 2001-12-02T22:00:00.000Z,
    "day": 1,
    "month": 1,
    "year": 2000,
  },
]
`;

exports[`Field static object 1`] = `
Object {
  "date": 2001-12-02T22:00:00.000Z,
  "day": 3,
  "month": 11,
  "year": 2001,
}
`;

exports[`Update shorcut: Field function array 1`] = `
Array [
  Object {
    "date": 2001-12-02T22:00:00.000Z,
    "day": 3,
    "month": 11,
    "year": 2001,
  },
]
`;

exports[`Update shorcut: Field function object 1`] = `
Object {
  "date": 2001-12-02T22:00:00.000Z,
  "day": 3,
  "month": 11,
  "year": 2001,
}
`;

exports[`Update shorcut: Field static array 1`] = `
Array [
  Object {
    "date": 2001-12-02T22:00:00.000Z,
    "day": 1,
    "month": 1,
    "year": 2000,
  },
]
`;

exports[`Update shorcut: Field static object 1`] = `
Object {
  "date": 2001-12-02T22:00:00.000Z,
  "day": 3,
  "month": 11,
  "year": 2001,
}
`;
